connectivity/drivers/ble/FEATURE_BLE/COMPONENT_BlueNRG_2/BlueNrg2HCIDriver.cpp

```diff
@@ -0,0 +1,631 @@
+ *     threads access the same SPI bus. Indeed it is common that the function
+ *     spiRead yield nothings even if the chip has signaled data with the irq
+ *     line. Sharing would make the situation worse and increase the risk of
+ *     timeout of HCI commands / response.
+ *   - This driver can be used even if the RTOS is disabled or not present it may
+ *     may be usefull for some targets.
+ *
+ * If The SPI is shared with other peripherals then the best option would be to
+ * handle SPI read in a real time thread woken up by an event flag.
+ *
+ * Other mechanisms might also be added in the future to handle data read as an
+ * event from the stack. This might not be the best solution for all BLE chip;
+ * especially this one.
+ */
+class TransportDriver : public CordioHCITransportDriver {
+public:
+    /**
+     * Construct the transport driver required by a BlueNRG_2 module.
+     * @param mosi Pin of the SPI mosi
+     * @param miso Pin of the SPI miso
+     * @param sclk Pin of the SPI clock
+     * @param irq Pin used by the module to signal data are available.
+     */
+    TransportDriver(PinName mosi, PinName miso, PinName sclk, PinName ncs, PinName irq)
+        : spi(mosi, miso, sclk), nCS(ncs), irq(irq), _spi_thread(osPriorityNormal, SPI_STACK_SIZE, _spi_thread_stack)
+    {
+        _spi_thread.start(mbed::callback(this, &TransportDriver::spi_read_cb));
+    }
+
+    virtual ~TransportDriver() { }
+
+    /**
+     * @see CordioHCITransportDriver::initialize
+     */
+    virtual void initialize()
+    {
+        // Setup the spi for 8 bit data, low clock polarity,
+        // 2-edge phase, with an 1MHz clock rate
+        spi.format(8, 1);
+        spi.frequency(1000000);
+
+        // Deselect the BlueNRG_2 chip by keeping its nCS signal high
+        nCS = 1;
+
+        wait_us(500);
+
+        // Set the interrupt handler for the device
+        irq.mode(PullDown); // set irq mode
+        irq.rise(mbed::callback(this, &TransportDriver::HCI_Isr));
+    }
+
+    /**
+     * @see CordioHCITransportDriver::terminate
+     */
+    virtual void terminate() { }
+
+    /**
+     * @see CordioHCITransportDriver::write
+     */
+    virtual uint16_t write(uint8_t type, uint16_t len, uint8_t *pData)
+    {
+        // repeat write until successfull. A number of attempt or timeout might
+        // be useful
+        while (spiWrite(type, pData, len) == 0) { }
+        return len;
+    }
+
+private:
+    uint16_t spiWrite(uint8_t type, const uint8_t *data, uint16_t data_length)
+    {
+        static const uint8_t header_master[] = { 0x0a, 0x00, 0x00, 0x00, 0x00 };
+        uint8_t header_slave[5];
+        uint16_t data_written = 0;
+        uint16_t write_buffer_size = 0;
+
+        _spi_mutex.lock();
+
+        irq.disable_irq();
+
+        /* CS reset */
+        nCS = 0;
+
+        // Wait until BlueNRG_2 is ready.
+        // When ready it will raise the IRQ pin.
+        _irq_timer.start();
+        while (!dataPresent()) {
```

Comment:
What is the average time to to raise the IRQ pin ?

Reply:
Should be checked with the team working on the BlueNRG-2 FW